File: advanced-programming-methods/labs/toy-interpreter-java/src/view/GUIStatementDialog.java
package view;

import model.statement.*;

import javax.swing.*;
import java.awt.event.*;

public class GUIStatementDialog extends JDialog
{
    private JPanel    contentPane;
    private JButton   buttonOK;
    private JButton   button_compound;
    private JButton   button_fork;
    private JButton   button_assign;
    private JButton   button_print;
    private JButton   button_if;
    private JLabel    label_value;
    private JButton   button_heap_write;
    private Statement statement;

    public GUIStatementDialog()
    {
        setContentPane(contentPane);
        setModal(true);
        setTitle("Get Statement");
        setLocationRelativeTo(null);
        getRootPane().setDefaultButton(buttonOK);
        pack();
        initHooks();

        setVisible(true);
    }

    /**
     * Getter for property 'statement'.
     *
     * @return Value for property 'statement'.
     */
    public Statement getStatement()
    {
        return statement;
    }

    /**
     * Setter for property 'statement'.
     *
     * @param statement Value to set for property 'statement'.
     */
    public void setStatement(Statement statement)
    {
        this.statement = statement;
    }

    private void initHooks()
    {
        buttonOK.addActionListener(new ActionListener()
        {
            public void actionPerformed(ActionEvent e)
            {
                onOK();
            }
        });
        // call onCancel() when cross is clicked
        setDefaultCloseOperation(DO_NOTHING_ON_CLOSE);
        addWindowListener(new WindowAdapter()
        {
            public void windowClosing(WindowEvent e)
            {
                onOK();
            }
        });

        // call onCancel() on ESCAPE
        contentPane.registerKeyboardAction(new ActionListener()
        {
            public void actionPerformed(ActionEvent e)
            {
                onOK();
            }
        }, KeyStroke.getKeyStroke(KeyEvent.VK_ESCAPE, 0), JComponent.WHEN_ANCESTOR_OF_FOCUSED_COMPONENT);


        button_compound.addActionListener(new ActionListener()
        {
            @Override
            public void actionPerformed(ActionEvent e)
            {
                System.out.println("Pressed compound");

                GUIStatementDialog first = new GUIStatementDialog();
                GUIStatementDialog second = new GUIStatementDialog();

                statement = new CompStatement(first.getStatement(), second.getStatement());
                label_value.setText(statement.toString());
            }
        });
        button_assign.addActionListener(new ActionListener()
        {
            @Override
            public void actionPerformed(ActionEvent e)
            {
                System.out.println("Pressed assign");

                String varname = GUI.getString(contentPane, "var name = ");
                GUIExpressionDialog dialog = new GUIExpressionDialog();

                statement = new AssignStatement(varname, dialog.getExpression());
                label_value.setText(statement.toString());
            }
        });
        button_print.addActionListener(new ActionListener()
        {
            @Override
            public void actionPerformed(ActionEvent e)
            {
                System.out.println("Pressed print");

                GUIExpressionDialog dialog = new GUIExpressionDialog();

                statement = new PrintStatement(dialog.getExpression());
                label_value.setText(statement.toString());
            }
        });
        button_if.addActionListener(new ActionListener()
        {
            @Override
            public void actionPerformed(ActionEvent e)
            {
                System.out.println("Pressed if");

                GUIExpressionDialog dialog1 = new GUIExpressionDialog();
                GUIStatementDialog dialog2 = new GUIStatementDialog();
                GUIStatementDialog dialog3 = new GUIStatementDialog();

                statement = new IfStatement(dialog1.getExpression(), dialog2.getStatement(), dialog3.getStatement());
                label_value.setText(statement.toString());
            }
        });
        button_fork.addActionListener(new ActionListener()
        {
            @Override
            public void actionPerformed(ActionEvent e)
            {
                System.out.println("Pressed fork");

                GUIStatementDialog dialog = new GUIStatementDialog();

                statement = new ForkStatement(dialog.getStatement());
                label_value.setText(statement.toString());
            }
        });
        button_heap_write.addActionListener(new ActionListener()
        {
            @Override
            public void actionPerformed(ActionEvent e)
            {
                System.out.println("Pressed heap write");

                String varname = GUI.getString(contentPane, "var name = ");
                GUIExpressionDialog expression = new GUIExpressionDialog();

                statement = new HeapWriteStatement(varname, expression.getExpression());
                label_value.setText(statement.toString());
            }
        });
    }

    private void onOK()
    {
        System.out.println("Pressed ok");

        if (statement == null)
        {
            GUI.showErrorDialog(contentPane, "Did not select any statement");
            return;
        }

        dispose();
    }
}

File: advanced-programming-methods/labs/toy-interpreter-sharp/view/GUIStatementDialog.cs
using System;
using Gtk;
using Repository;
using Controller;
using Model;
using View;

namespace View
{
    public partial class GUIStatementDialog : Dialog
    {
        private Statement statement;

        public GUIStatementDialog()
        {
            Title = "Get Statement";
            this.Build();

        }

        public Statement Statement
        {
            get
            {
                return statement;
            }

            set
            {
                statement = value;
            }
        }

        protected void OnClose(object sender, EventArgs e)
        {
            cleanup();
        }

        protected void OnDeleteEvent(object sender, EventArgs e)
        {
            cleanup();

            var args = (DeleteEventArgs)e;
            args.RetVal = true; // prevent close
        }

        protected void OnButtonOkClicked(object sender, EventArgs e)
        {
            cleanup();
        }

        protected void cleanup()
        {
            Console.WriteLine("Pressed ok");

            if (statement == null)
            {
                GUI.ShowErrorDialog(this, "Did not select any statement");
                return;
            }

            HideAll();
        }

        protected void OnButtonCompoundClicked(object sender, EventArgs e)
        {
            Console.WriteLine("Pressed compound");

            statement = new CompStatement(GUI.GetStatement(), GUI.GetStatement());
            label_value.Text = statement.ToString();
        }

        protected void OnButtonAssignClicked(object sender, EventArgs e)
        {
            Console.WriteLine("Pressed assign");


            string varname = GUI.GetString(this, "var name = ");
            statement = new AssignStatement(varname, GUI.GetExpression());
            label_value.Text = statement.ToString();
        }

        protected void OnButtonPrintClicked(object sender, EventArgs e)
        {
            Console.WriteLine("Pressed print");

            statement = new PrintStatement(GUI.GetExpression());
            label_value.Text = statement.ToString();
        }

        protected void OnButtonIfClicked(object sender, EventArgs e)
        {
            Console.WriteLine("Pressed If");

            statement = new IfStatement(GUI.GetExpression(), GUI.GetStatement(), GUI.GetStatement());
            label_value.Text = statement.ToString();
        }

        protected void OnButtonForkClicked(object sender, EventArgs e)
        {
            Console.WriteLine("Pressed Fork");

            statement = new ForkStatement(GUI.GetStatement());
            label_value.Text = statement.ToString();
        }

        protected void OnButtonLockEnterClicked(object sender, EventArgs e)
        {
            Console.WriteLine("Pressed Lock Enter");

            statement = new LockEnterStatement(GUI.GetInt(this, "lock number ="));
            label_value.Text = statement.ToString();
        }

        protected void OnButtonLockExitClicked(object sender, EventArgs e)
        {
            Console.WriteLine("Pressed Lock Exit");

            statement = new LockExitStatement(GUI.GetInt(this, "lock number ="));
            label_value.Text = statement.ToString();
        }
    }
}
